packages/flutter/lib/src/cupertino/sheet.dart

```diff
@@ -0,0 +1,414 @@
+/// the content on the [CupertinoSheetRoute].
+///
+/// `useNestedNavigation` controls whether or not boilerplate code is setup for
+/// enabling nested navigation for the sheet.
+///
+/// When `useNestedNavigation` is set to `true`, any route pushed to the stack
+/// from within the context of the [CupertinoSheetRoute] will display within that
+/// sheet. System back gestures, and programatic pops on the initial route in a
+/// sheet will also be intercepted to pop the whole [CupertinoSheetRoute]. If
+/// another [Navigator] setup is needed, like for example to enamble named routes
+/// or the pages API, then it is recommended to directly push a [CupertinoSheetRoute]
+/// to the stack with whatever configuration needed. See the implementation below
+/// for the boilerplate provided by `showCupertinoSheet` to get started.
+///
+/// The whole sheet can be popped at once by either dragging down on the shet,
+/// or calling [CupertinoSheetRoute.popSheet].
+///
+/// iOS sheet widgets are generally designed to be tightly coupled to the context
+/// of the widget that opened the sheet. As such, it is not recommended to directly
+/// link to a route outside of the sheet, without first popping the sheet. If that
+/// is needed however, it can be done by pushing to the root Navigator.
+///
+/// If `useNestedNavigation` is left as `false`, then a [CupertinoSheetRoute]
+/// will be shown with no [Navigator] widget. Multiple calls to `showCupertinoSheet`
+/// can still be made to show multiple stacked sheets, if desired.
+///
+/// In both cases, new [CupertinoSheetRoute]s will be pushed to the root
+/// [Navigator]. This is to ensure the previous routes animate correctly.
+///
+/// Returns a [Future] that resolves to the value (if any) that was passed to
+/// [Navigator.pop] when the sheet was closed.
+///
+/// {@tool dartpad}
+/// This example shows how to navigate to use [showCupertinoSheet] to display a
+/// Cupertino sheet widget with nested navigation.
+///
+/// ** See code in examples/api/lib/cupertino/sheet/cupertino_sheet.2.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [CupertinoSheetRoute] the basic route version of the sheet view.
+///  * [showCupertinoDialog] which displays an iOS-styled dialog.
+Future<T?> showCupertinoSheet<T>({
+  required BuildContext context,
+  required WidgetBuilder pageBuilder,
+  bool useNestedNavigation = false,
+}) {
+  final WidgetBuilder builder;
+  if (!useNestedNavigation) {
+    builder = pageBuilder;
+  } else {
+    builder = (BuildContext context) {
+      return NavigatorPopHandler(
+        onPopWithResult: (T? result) {
+          Navigator.of(context, rootNavigator: true).maybePop(result);
+        },
+        child: Navigator(
+          initialRoute: '/',
+          onGenerateRoute: (RouteSettings settings) {
+            return CupertinoPageRoute<void>(
+              builder: (BuildContext context) {
+                return PopScope(
+                  canPop: settings.name != '/',
```

Comment:
I think what this does now as-written is to prevent programmatically popping the nested Navigator when on route '/'. So I would double check that that is necessary.

Reply:
`onGenerateInitialRoutes` makes sense. Is using both `PopScope` and `NavigatorPopHandler`? Trying to cover both system back gestures and programatic pops.

Reply:
using popScope and NavigatorPopHandler both sounds good, can you add test for the 2 scenarios?

1. calling Navigator.of(context).pop can pop sheet with nestNavigator
2. using system back gesture on sheet with nestnavigator that has multiple sub routes pops each route correctly and finally pops the entire sheet at the end?